design/baremetal-operator/sub-states.md

```diff
@@ -0,0 +1,184 @@
+
+## Design Details
+
+* The `ProvisionStatus` structure is extended with a new string field
+  `CurrentActivity` and a new integer field `Progress`.
+* The `Progress` value is a number from 0 to 100 that corresponds to estimated
+  procentage of the work that has been done so far.
+* The values of the fields are re-calculated based on the result described
+  below on each reconciliation.
+
+### Ironic steps basics
+
+This section explains the basics of Ironic clean/deploy steps. Feel free to
+skip it if you're familiar with Ironic.
+
+Ironic has three major processes that are of interest for us:
+
+* *Inspection* is a process of populating hardware information.
+* *Cleaning* is a process of preparing a node for being usable for deployment.
+* *Deployment* is a process of provisioning of a node.
+
+All three processes usually (with a few exceptions that are skipped for
+brevity) work by booting a ramdisk on the target node and executing actions
+from within it. When *fast track* mode is on, the ramdisk is booted only once
+and inspection, cleaning and deployment happen in a sequence.
+
+Cleaning and deployment processed are split into potentially pluggable *clean*
+or *deploy steps*. The said steps can be defined both inside Ironic itself or
+inside the ramdisk. When a process is running, the current step is stored in
+the Node's `clean_step` or `deploy_step` fields. All steps that are currently
+scheduled for execution are cached in
+`Node.driver_internal_info["clean_steps"]` or
+`Node.driver_internal_info["deploy_steps"]`.
+
+**Warning:** deploy (and in the future also automated clean) steps are
+populated on demand. Before the ramdisk is actually booted, not all steps may
+be in the cache.
+
+### Implementation Details/Notes/Constraints
+
+The new fields will be populated based on the following rules, where `Node` is
+an Ironic Node object. The rules are defined in pseudo-code.
+
+The calculations of the progress are based on the assumption that automated
+cleaning is enabled. It starts with the value of 0.
+
+* IF `Node.provision_state` in "cleaning", "clean wait":
+
+  * IF `Node.clean_step` is null:
+
+    Activity is `Preparing to configure host`.
+
+  * IF `Node.clean_step["interface"]` == "raid":
+
+    Activity is `Configuring RAID`.
+
+  * IF `Node.clean_step["interface"]` == "bios":
+
+    Activity is `Configuring firmware settings`.
+
+  * IF `Node.clean_step["step"]` == "erase\_devices\_metadata":
+
+    Activity is `Deleting disk metadata`.
+
+  * Progress is determined by looking at the clean steps cache and determining
+    which of the step is currently executed, assuming that cleaning takes 30%
+    of the total time. For example, if one and only step is being executed,
+    Progress is 15.
+
+* IF `Node.provision_state` in "deploying", "deploy call-back":
+
+  * IF `Node.deploy_step` is null OR `Node.deploy_step["step"]` is "deploy":
+
+    Activity is `Preparing to write image`. Progress is 30.
```

Comment:
These progress values would apply after each step is completed, right? So when we move from preparing-to-write to writing, we would set the progress to 30?

Reply:
No, I'm trying to write an algorithm here, so when this condition is true, progress is set to 30.

Reply:
OK, that's a little confusing. What is happening before we get to preparing-to-write that accounts for 30 percent of the total work? Booting the agent?

Reply:
The whole automated cleaning (since for us it's a part of the deployment).